spatial_cell_cpu.hpp

```diff
@@ -162,30 +153,151 @@ namespace spatial_cell {
+         RHO_V = other.RHO_V;
+         RHOLOSSADJUST = other.RHOLOSSADJUST;
+         velocityBlockMinValue = other.velocityBlockMinValue;
+         ACCSUBCYCLES = other.ACCSUBCYCLES;
+         N_blocks = other.N_blocks;
+         for (uint i=0; i<2; ++i) {
+            max_dt[i] = other.max_dt[i];
+         }
+         for (uint i=0; i<3; ++i) {
+            V[i] = other.V[i];
+            V_R[i] = other.V_R[i];
+            V_V[i] = other.V_V[i];
+            P[i] = other.P[i];
+            P_R[i] = other.P_R[i];
+            P_V[i] = other.P_V[i];
+         }
+         return *this;
+      }
+      void ResizeClear(const uint newSize) {
+         // Resizes the vmesh localToGlobalMap, clears the vmesh GlobalToLocalMap,
+         // and resizes the velocity block container.
+         // Contents of the localToGlobalMap or the VBC are not edited.
+         vmesh->setNewSize(newSize);
+         vmesh->clearMap(newSize);
+         blockContainer->setNewSize(newSize);
+      }
+      void Scale(creal factor) {
+         RHO *= factor;
+         RHO_R *= factor;
+         RHO_V *= factor;
+         for (uint i=0; i<3; ++i) {
+            P[i] *= factor;
+            P_R[i] *= factor;
+            P_V[i] *= factor;
+         }
+         // Now loop over whole velocity space and scale the values
+         for (vmesh::LocalID blockLID=0; blockLID < vmesh->size(); ++blockLID) {
+            // Pointer to target block data
+            Realf* toData = blockContainer->getData(blockLID);
+            // Scale data
+            for (uint i=0; i<WID3; ++i) {
+               toData[i] = toData[i] * factor;
+            }
+         }
+      }
+      void Increment(const Population& other, creal factor) {
+         // Note: moments will be invalidated.
+         // Loop over the whole velocity space, and add scaled values.
+         for (vmesh::LocalID incBlockLID=0; incBlockLID<(other.vmesh)->size(); ++incBlockLID) {
+            const Realf* fromData = (other.blockContainer)->getData(incBlockLID);
+
+            // Global ID of the block containing incoming data
+            vmesh::GlobalID incBlockGID = (other.vmesh)->getGlobalID(incBlockLID);
+
+            // Get local ID of the target block. If the block doesn't exist, create it.
+            vmesh::GlobalID toBlockLID = vmesh->getLocalID(incBlockGID);
+            if (toBlockLID == vmesh->invalidLocalID()) {
+               bool success = vmesh->push_back(incBlockGID);
+               toBlockLID = blockContainer->push_back_and_zero();
+               Real* parameters = blockContainer->getParameters(toBlockLID);
+               vmesh->getBlockInfo(incBlockGID, parameters+BlockParams::VXCRD);
+            }
+
+            // Pointer to target block data
+            Realf* toData = blockContainer->getData(toBlockLID);
+            // Add scaled values from source cells
+            for (uint i=0; i<WID3; ++i) {
+               toData[i] += fromData[i] * factor;
+            }
```

Comment:
Can't these be somewhat parallelised? Maybe vectorised, if they get called inside threaded regions?

Reply:
Feel free to implement!
(Modern compilers may be able to auto-vectorize this anyway)